gradle: Drop init-deps.gradle script

britter opened the pull request:
Instead of injecting a tasks into all projects, the gradle build support now runs `--write-verification-metadata` which according to [this forum post](https://discuss.gradle.org/t/force-resolving-and-downloading-all-dependencies/35942/5) is a way of resolving all dependencies that have been declared correctly.

For projects that have dependency verification turned off, we run `--write-verification-metadata sha256` as this is enough to force a download of all dependencies in the project. Projects that have dependency verification turned on require to run `--write-verification-metadata pgp,sha256`. This is because the `pgp` will also force a download of asc signature files, which are required at dependency verification when the build runs later. Without the `pgp` flag the signature files would be missing in the lock file which results in the build failing due to dependency verification.

I ran the dependency update for all Gradle packages that I could find via grep, and rebuilt the ones that had changes in the dependency file. Lockfiles for most packages did not change which indicates that this approach works.

Resolves #381969 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

britter committed:
gradle: Replace init-deps.gradle with --write-verification-metadata

It does effectively the same but is configuration cache and project
isolation compatible.

By default to capture dependencies, the setup-hook runs `gradle help
--write-verification-metadata sha256` which will cause Gradle to
download all dependencies. In case a project has dependency verification
turned on, it runs with the `pgp` flag in addition, so Gradle also downloads
signatures that are required later by the sandboxed build. Since this
results in requests to key servers, mitm-cache is configured to ignore
key server requests.

Fixes #381969

britter committed:
atlauncher: update dependencies after changes to gradle build support

This change is due to the fact that the atlauncher build uses the
jitpack.io registry which makes the latest git commit of a project
available as a maven dependency.

britter committed:
jadx: update dependencies after changes to gradle build support

britter committed:
libeufin: update dependencies after changes to gradle build support

britter committed:
slimevr-server: update dependencies after changes to gradle build support

britter committed:
json2cdn: update dependencies after changes to gradle build support

britter committed:
mucommander: update dependencies after changes to gradle build support

britter committed:
mindustry: update dependencies after changes to gradle build support

britter commented:
Optimistically pinging a few people to get to the bottom of the build failure of `pkgs.freenet`: @FliegendeWurst @chayleaf @Infinidoge @liff @lorenzleutgeb @nagy

liff commented:
Looks like the dependency to `bcpkix-jdk15on` 1.48 might originate from the usage of the [Android plugin](https://github.com/hyphanet/fred/blob/f937e1a4c966fbc8e07f68051b4395ade4a5c852/build.gradle#L7-L9). The version is resolved to [0.9.2](https://mvnrepository.com/artifact/com.android.tools.build/gradle/0.9.2), which depends on [builder:0.9.2](https://mvnrepository.com/artifact/com.android.tools.build/builder/0.9.2), which in turn depends on bcpkix-jdk15on 1.48.

britter commented:
@liff you're right! I somehow forgot about build dependencies 🤦🏻‍♂️ But here it is: https://scans.gradle.com/s/p6b4e4uxqpjhc/build-dependencies?dependencies=bcprov&expandAll&focusedDependency=WzAsMCw0NSxbMCwwLFswXV1d&focusedDependencyView=versions&toggled=W1swLDBdLFswLDBdXQ
So now the question is why does the presence of this dependency cause the build to fail when the dependency lock file is generated using the new way.

britter commented:
Okay, I'm making progress. The reason it's failing for freenet is that fact that freenet has Gradle dependency verification turned on, including signature verification. I modified the setup hook to use `--write-verification-metadata sha256` which will only download artifacts and calculate their checksums locally. It will not download the signatures. That's why we're seeing removals of `.asc` entries in the diff I shared above.

Now in order to fix this I tried running `--write-verification-metadata pgp,sha256`. After this the the diff looks different. The `asc` entries are preserved. But there's a new problem. It now adds entries like the following at the top:

```json
"https:/": {
  "keys.openpgp": {
   "org/pks/lookup": "sha256-cevzjs9C1E3OhGqXcCHElP8SsEc3vmJgZAP7a+opfpk="
  },
  "keyserver.ubuntu": {
   "com/pks/lookup": "sha256-PwKyEeHMflcIm4ZiB9GhbI/JEgnzYKkNbnvyDZJrhcQ="
  },
  "pgp.mit": {
   "edu/pks/lookup": "sha256-RObs2mIBKF5DtQ1F3OYV+rzRfP7w/jJrlBgNWnTfe6k="
  },
```

This looks like the requests Gradle is making to the key servers in order to download signing keys. If this is included in the lock file, building the package fails because it tries to "download" https://keys.openpgp.org/pkgs/lookup which results in 404.

One way of fixing this would be using `jq` at the end of the generated `fetch-dep.sh` script to remove the entries ending with `pkgs/lookup`
again.

WDYT?

britter commented:
Okay, I added some code to remove the keyserver requests from the dependency lock file again and made sure signatures are only added when required. So this is ready for review now, from my  POV.

kira-bruneau reviewed (approved):
Thanks for working on this @britter! :heart: This also solved a problem I was having building some minecraft mods with Nix!

The changes look good to me, but I'll give some time for @chayleaf to respond before merging this.

chayleaf commented on doc/languages-frameworks/gradle.section.md:
```diff
@@ -163,7 +163,7 @@ The Gradle setup hook accepts the following environment variables:
   the package if `doCheck` is set to `true`. Defaults to `test`.
 - `gradleUpdateTask` - the Gradle task (or tasks) to be used for
```
This being named `gradleUpdateTask` makes no sense.

britter replied:
Why? Because the command line does not explicitly run a Gradle task? FWIW Gradle implicitly `runs` help.

Do you have a proposal for a better name?

britter replied:
I'm confused by your comment. This is the task that users can specify in order to capture dependencies. Why do you think `gradleUpdateTask` is not appropriate anymore?

chayleaf replied:
Because `--write-verification-metadata` isn't a task, it's a CLI flag

britter replied:
@chayleaf @kira-bruneau I'd like to clarify the behavior of `--write-dependency-metadata`. It will resolve any dependencies, no matter which tasks are requested on the command line. That's why you usually run it without specifying a task, which will then run the default task, which is `help` if you've not configured to be something else. I'm strongly against running `$gradleBuildTask` and/or `$gradleCheckTask` because it will take way longer.

I think it makes sense to make sure it runs help, in case a project has the default task configured to something expensive. If we change the default of `gradleUpdateTask` to `help --write-dependency-metadata sha256` it makes sense again to call it `gradleUpdateTask`.

kira-bruneau replied:
Ok thanks for clearing that up! I think it's fair to mark this as resolved with the explicit help task included.

chayleaf commented on pkgs/development/tools/build-managers/gradle/init-deps.gradle:
```diff
@@ -1,10 +0,0 @@
-gradle.projectsLoaded {
```
It should be replaced with a task that indicates the migration steps for out-of-tree projects. Update scripts require usually require a bit of manual work anyway so it's fine to break backwards compatibility right away, but it should be explained.

britter replied:
Can you elaborate some more what kind of problem you see for out-of-tree projects? As far as I can tell the Gradle build support does not give access to the init script directly. How would an out-of-tree project be using this?

chayleaf replied:
The script used to expose the `nixDownloadDeps` task, which would be used by any project that needs to invoke Gradle multiple times to fetch all necessary deps. Since this task has been removed, this is a breaking change for update scripts.

kira-bruneau replied:
> Can you elaborate some more what kind of problem you see for out-of-tree projects? As far as I can tell the Gradle build support does not give access to the init script directly. How would an out-of-tree project be using this?

The changes you had to make to jabref is good example of what @chayleaf is describing.

You had the context to know `gradle nixDownloadDeps` should've been replaced with `gradle --write-verification-metadata`, but anyone who's still calling `gradle nixDownloadDeps` out of tree will now unexpectedly have their builds fail without a good explanation of how to fix it.

britter replied:
Fair enough. I'll add it back again. I thought this is what people would expect when using `unstable`. What is the timeline for completely removing it after this PR is merged?

liff replied:
Should the `nixDownloadDeps` task also come with a deprecation notice? How about an entry in the [release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)?

kira-bruneau replied:
> I thought this is what people would expect when using unstable

This PR won't just affect unstable though since `nixDownloadDeps` was included in 24.11, and there should be a clear migration path for the next release.

It would be ok to make this change without a migration if `nixDownloadDeps` didn't already exist in a stable release.

> What is the timeline for completely removing it after this PR is merged?

24.11 will reach EOL once 25.05 is released, so it would be ok fully remove it after 25.05 is out.

> Should the nixDownloadDeps task also come with a deprecation notice? How about an entry in the [release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md)?

Oh yep good point - an entry should also be added to the release notes indicating the breaking change.

chayleaf commented on pkgs/development/tools/build-managers/gradle/update-deps.nix:
```diff
@@ -143,6 +142,9 @@ lib.makeOverridable (
         NIX_BUILD_SHELL=${runtimeShell} nix-shell --pure --run ${gradleScript} ${nixShellKeep} ${sourceDrvPath}
       fi${lib.optionalString silent " >&2"}
       kill -s SIGINT "$MITM_CACHE_PID"
+      # Remove entries for requests Gradle made to keyservers in order to look up signing keys
```
can this be a regex flag to mitm-cache?

britter replied:
Okay, would you mind creating a release of mitm-cache? If it's okay, I'd like to replace the `jq` code with `--forget` in a separate PR.

chayleaf replied:
Done, a tag and a release are now published.

britter replied:
Updated mitm-cache to 1.1.2 and used the new forget flag.

kira-bruneau replied:
Hmm it looks like It's not clear why, but I'm getting multiple entries with the same key on the latest update:

```
> nix build .#jabref.mitmCache.updateScript && ./result
FAILURE: Build failed with an exception.

* What went wrong:
Multiple entries with same key: com.tobiasdiez:easybind:2.2.1-SNAPSHOT=org.gradle.api.internal.artifacts.verification.model.ImmutableComponentVerificationMetadata@2fd1d69b and com.tobiasdiez:easybind:2.2.1-SNAPSHOT=org.gradle.api.internal.artifacts.verification.model.ImmutableComponentVerificationMetadata@fb625c0

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.12.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 17s
3 actionable tasks: 1 executed, 2 up-to-date
```

britter replied:
@kira-bruneau I've looked into this all morning and I think you've discovered a bug in Gradle. I've tried to find a way to work around it but without success. I've reported this back to the Gradle team here: https://github.com/gradle/gradle/issues/32739. I think we've to put this work on hold until I hear back from them unfortunately.

chayleaf reviewed (changes requested):
If `freenet` is fixed, this LGTM other than some minor stuff.

chayleaf commented on pkgs/development/tools/build-managers/gradle/setup-hook.sh:
```diff
@@ -54,7 +54,11 @@ gradleUpdateScript() {
+    if [ -f gradle/verification-metadata.xml ]; then
+        verificationFlags="pgp,sha256"
+    fi
+    gradle ${enableParallelUpdating:+--parallel} ${gradleUpdateTask:---write-verification-metadata $verificationFlags}
```
`defaultTasks` may not necessarily match what the derivation will run `$gradleBuildTask` and `$gradleCheckTask` should be used here instead of the default task

britter replied:
`--write-verification-metadata` will resolve any dependencies. There's no need to run specific tasks as far as I can tell.

britter replied:
I should say any, but the dependencies that are loaded dynamically at task execution time using detached configurations. But I think it's fair to keep it this way because it's fast then running check and build. It will work for most projects - it works for the projects in nixpkgs. For those that make use of detached configurations, they can still override `gradleUpdateTask`.

chayleaf replied:
No, [Gradle docs](https://docs.gradle.org/6.4/userguide/dependency_verification.html) explicitly mention that the default task is used by default:

> 

> There are dependencies that Gradle *cannot* discover this way. In particular, you will notice that the CLI above uses the `help` task. If you don’t specify any task, **Gradle will automatically run the default task** and generate a configuration file at the end of the build too.

So it makes sense to use the user-provided tasks instead of the default one.

britter replied:
See my comment [here](https://github.com/NixOS/nixpkgs/pull/383115#discussion_r1973169150). I've changes the code to explicitly run `help`.

kira-bruneau replied:
It's not clear why, but it looks like the latest commit doesn't handle the duplicate entries like the original commit did:

```
> nix build .#jabref.mitmCache.updateScript && ./result
FAILURE: Build failed with an exception.

* What went wrong:
Multiple entries with same key: com.tobiasdiez:easybind:2.2.1-SNAPSHOT=org.gradle.api.internal.artifacts.verification.model.ImmutableComponentVerificationMetadata@2fd1d69b and com.tobiasdiez:easybind:2.2.1-SNAPSHOT=org.gradle.api.internal.artifacts.verification.model.ImmutableComponentVerificationMetadata@fb625c0

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.12.1/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 17s
3 actionable tasks: 1 executed, 2 up-to-date
```

chayleaf commented on pkgs/development/tools/build-managers/gradle/setup-hook.sh:
```diff
@@ -54,7 +54,11 @@ gradleUpdateScript() {
     runHook preGradleUpdate
 
-    gradle ${enableParallelUpdating:+--parallel} ${gradleUpdateTask:-nixDownloadDeps}
+    local verificationFlags=sha256
```
there should probably be a small comment explaining that this is for fetching all dependencies and not for integrity checking (and is there any reason to use `pgp,sha256` here?)

britter commented:
@chayleaf 

> If `freenet` is fixed, this LGTM other than some minor stuff.

Yes after I added the jq filtering, the lockfile for `freenet` does not change when running the dependency update using the new way. So this is resolved.

I added a comment to explain why we're using dependency verification. Please take a look at my other comments/questions.

Thanks!

wentasah reviewed (approved):
`fastddsgen` still works, so I'm not against :-)

kira-bruneau reviewed (changes requested):
@chayleaf found some issues that should still be addressed before this is merged.

britter committed:
mitm-cache: 0.1.1 -> 0.1.2

britter commented:
@chayleaf @kira-bruneau please take a look again.

kira-bruneau commented on pkgs/applications/office/jabref/default.nix:
```diff
@@ -125,8 +125,8 @@ stdenv.mkDerivation rec {
 
-    gradle nixDownloadDeps -Dos.arch=amd64
-    gradle nixDownloadDeps -Dos.arch=aarch64
+    gradle --write-verification-metadata sha256 -Dos.arch=amd64
```
Would it also make sense to use the help task for these explicit calls to gradle?

britter commented:
This is currently blocked by https://github.com/gradle/gradle/issues/32739
